### PR TITLE
`enable_model_cpu_offload`

### PR DESCRIPTION
**Pull request by pcuenca**
It's surprisingly more involved than expected, see comments below.

### CODE REVIEW

File: src/diffusers/models/autoencoder_kl.py [hunk 1]
```diff
@@ -183,3 +183,12 @@ def forward(
...
         return DecoderOutput(sample=dec)
+
+
+class VAEDecoder(nn.Module):
+    def __init__(self, vae: AutoencoderKL):
+        super().__init__()
+        self.vae = vae
+
+    def forward(self, z: torch.FloatTensor, return_dict: bool = True):
+        return self.vae.decode(z, return_dict=return_dict)
```
Comment (pcuenca) [inline src/diffusers/models/autoencoder_kl.py:194]: New module to trigger the previous hook, because it happens in the `pre_forward` method.

Unfortunately, I think we can't register `AutoencoderKL.decoder` instead, because the decode method performs a convolution before invoking the decoder: https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/autoencoder_kl.py#L123

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 2]
```diff
@@ -161,6 +161,7 @@ def __init__(
             feature_extractor=feature_extractor,
         )
         self.vae_scale_factor = 2 ** (len(self.vae.config.block_out_channels) - 1)
+        self.decoder = VAEDecoder(self.vae)
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:164]: We use the new `decoder` instead of the `vae` for decoding.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 3]
```diff
@@ -192,20 +195,47 @@ def enable_sequential_cpu_offload(self, gpu_id=0):
 
         device = torch.device(f"cuda:{gpu_id}")
 
-        for cpu_offloaded_model in [self.unet, self.text_encoder, self.vae]:
+        for cpu_offloaded_model in [self.unet, self.text_encoder, self.decoder]:
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:198]: And we need to register the `decoder` instead of the `vae`.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 4]
```diff
@@ -192,20 +195,47 @@ def enable_sequential_cpu_offload(self, gpu_id=0):
...
+        for cpu_offloaded_model in [self.text_encoder, self.unet, self.decoder]:
+            _, hook = cpu_offload_with_hook(cpu_offloaded_model, device, prev_module_hook=hook)
+
+        if self.safety_checker is not None:
+            _, hook = cpu_offload_with_hook(self.safety_checker, device, prev_module_hook=hook)
+
+        # We'll offload the last model manually.
+        # Alternative: daisy-chain with the first one so a new inference will offload the last model,
+        # but then we need access to `hook.hook.prev_module_hook`.
+        self.offload_hook = hook
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:229]: I can't think of a simple and automatic way to do this.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 5]
```diff
@@ -192,20 +195,47 @@ def enable_sequential_cpu_offload(self, gpu_id=0):
...
+
     @property
     def _execution_device(self):
         r"""
         Returns the device on which the pipeline's models will be executed. After calling
         `pipeline.enable_sequential_cpu_offload()` the execution device can only be inferred from Accelerate's module
         hooks.
         """
-        if self.device != torch.device("meta") or not hasattr(self.unet, "_hf_hook"):
+        if not hasattr(self.unet, "_hf_hook"):
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:238]: Device is no longer `meta` when using `enable_model_offload`. I'm not fully sure why we need it for `enable_sequential_cpu_offload`, as `_hf_hook` seems to be enough (testing neeeded).

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 6]
```diff
@@ -366,7 +396,7 @@ def run_safety_checker(self, image, device, dtype):
 
     def decode_latents(self, latents):
         latents = 1 / self.vae.config.scaling_factor * latents
-        image = self.vae.decode(latents).sample
+        image = self.decoder(latents).sample
```
Comment (pcuenca) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:399]: As explained above, so `forward` is invoked and the hook triggers.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 4 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:229]: Think that's fine, would maybe just call it `self.final_offload_hook`

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:399]: Instead of doing this change could we maybe just retrieve the hook from `self.vae` and call it manually. E.g. 

```py
if hasattr(self.vae, "_hf_hook") and hasattr("self.vae._hf_hook", "prev_model_hook"):
    # offload unet
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py [hunk 7]
```diff
@@ -647,6 +677,10 @@ def __call__(
             # 9. Run safety checker
             image, has_nsfw_concept = self.run_safety_checker(image, device, prompt_embeds.dtype)
 
+        # Offload last model to CPU
+        if hasattr(self, "offload_hook") and self.offload_hook is not None:
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:681]: Think that's totally fine :-)

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:399]: That would offload the unet, but will not move the vae to cuda, because we are not using the vae's `forward`.

Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:399]: Ah yeah good point. Let's maybe also move it manually then there?

Comment (pcuenca) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:399]: I hate it too, but doing it manually is going to be ugly, I think. I'll try it.

**Comment by keturn** [comment]
https://github.com/huggingface/diffusers/blob/3f6a7b99f4d4d93d55be1d5e5e3835da180c705d/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L225-L229

It might be too much to assume the sequence the models will be used in.

For example, with img2img you'll be doing vae.encode before the unet. Is that vae.encode before or after the text_encoder? Who knows! Sure, a specific pipeline implementation might know its own design, but if you're building a base for other application code, you don't know what that order is. Especially now that the pipeline has the option of taking `prompt_embeds` instead of encoding the prompt itself.

And not that I would recommend it in combination with this offloading technique, but callbacks might also be doing vae.decode on the work in progress before control goes back to the pipeline for the next unet step.

Potential alternative:

Have some resource that tracks the "current model" (or "most recently loaded model" or however you want to phrase it). Then you can do

```py
if device.current_model != this_model:
    device.offload()
    device.load(this_model)
```

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 5 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py:243]: LGTM

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (keturn) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py]: Another approach, which would come with its own set of difficulties, would be to teach accelerate.add_hook_to_module to hook arbitrary methods instead of hardcoding `forward`. Then ask it to put the same hook around `forward` and `decode`.

But the fact that AutoencoderKL is _already_ two separate Encoder and Decoder modules under the hood makes me wonder if there's a better way.

**Comment by patrickvonplaten** [comment]
> loaded

I'm not sure how we could implement something like this as it would mean that something is constantly tracking which model is used which is not really possible. 

If people want to implement this feature they have to know the computation graph of the pipeline, otherwise things become to complex, magic under the hood IMO.

File: src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py]: I like this idea! Would be cool if we could put up a feature request for accelerate :-). This would indeed help a lot. Wdyt @pcuenca ?

**Review by patrickvonplaten — APPROVED** [review]
In general, I think this is a very nice addition and should give us a nice trade-off between speed and memory. 

Would be cool to see some numbers here. 
I understand @keturn's point that it's not straight-forward to know what module ran before what other module, but I don't really think there is a way around it and if there is, things become to magic. Also one could argue that computation flow is something that PyTorch users are relatively well educated in.

I second @keturn's idea of trying to introduce a keyword in accelerate that let's one override "forward" as the method that triggers offload and onload. @pcuenca what are yoru thoughts here?

**Comment by keturn** [comment]
> I'm not sure how we could implement something like this as it would mean that something is constantly tracking which model is used which is not really possible.

As you're already installing model hooks, it's straightforward enough for that pre-forward hook to send a "this is the model being used" message.

> If people want to implement this feature they have to know the computation graph of the pipeline, otherwise things become too complex, magic under the hood IMO.

You're not wrong, but on the other hand isn't this the selling point of `accelerate`? Do magic under the hood so models and their data are on the right devices when you need them, without adding explicit device management code to your application?

**Comment by pcuenca** [comment]
> I second @keturn's idea of trying to introduce a keyword in accelerate that let's one override "forward" as the method that triggers offload and onload. @pcuenca what are yoru thoughts here?

Interesting idea! But this is starting to feel like maybe a bit too much complexity, given that it wouldn't be that hard to offload/onload the models ourselves, in this particular case. Doing it recursively as in the classic `enable_sequential_cpu_offload` is complicated and finicky, so it's certainly helpful to use a battle-tested library that is well tested and proven to work. I'm not sure if "sequential chained top-level modules that don't always use forward" is general enough to be abstracted.

I'm willing to listen to opinions and usage examples, and happy to work on PRs (in `accelerate` or elsewhere) to make this happen. Just trying to make sure we are not inadvertently over-engineering a solution.

In this case, we would need to hook the `decode` method, so our simple hook loop needs to be unrolled because we have different cases:

```python
_, hook = cpu_offload_with_hook(self.text_encoder, device)
_, hook = cpu_offload_with_hook(self.unet, device, prev_module_hook=hook)
_, hook = cpu_offload_with_hook(self.vae, device, prev_module_hook=hook, wrapped_method="decode")

if self.safety_checker is not None:
    _, hook = cpu_offload_with_hook(self.safety_checker, device, prev_module_hook=hook)
```

But if we were to use the image2image pipelines, as @keturn also mentioned, then we'd need to also squeeze in the `encode` method somewhere.

**Comment by keturn** [comment]
Alternate proof of concept here: https://github.com/invoke-ai/InvokeAI/blob/f39c806aa6b33f637b7f7fecc2af4bd041e5fde7/ldm/invoke/offloading.py#L11

The problems I've run in to with that so far are:
- using torch's built-in hooks seemed like a good idea, but unlike accelerate's hooks, these are blind to kwargs.
- still have the `vae.decode` problem, because torch's hooks still only apply to the forward method.
- can't really use `model.device` for anything, because it is probably `cpu` now but that doesn't mean it will be `cpu` when it actually executes.

I was thinking we're stuck with the `model.device` problem until I remembered that it's not a standard torch.nn.Module property; `model.device` is a diffusers.ModelMixin invention. hmm. 🤔

**Comment by keturn** [comment]
Another other approach for the `vae.decode` problem:

Instead of teaching the hooking-things to hook arbitrary methods, teach the methods how to use the existing hooks. i.e.

```py
class Autoencoder(ModelMixin):
    @use_forward_hooks
    def decode(self, x):
        ...
```

It's not perfect, because to implement `use_forward_hooks` you'd have to dig around in objects for state that there's no public API for (torch's `_forward_pre_hooks` and/or accelerate's `_hf_hook`). But we're pretty much ready to resort to that anyway.

**Comment by pcuenca** [comment]
> ```python
> class Autoencoder(ModelMixin):
>     @use_forward_hooks
>     def decode(self, x):
>         ...
> ```

This looks promising. But it would be built on top of the proposed addition to accelerate to work on arbitrary methods, right? Or do you mean to look for the hooks associated to `forward` in the object and manipulate them?

**Comment by patrickvonplaten** [comment]
> Another other approach for the `vae.decode` problem:
> 
> Instead of teaching the hooking-things to hook arbitrary methods, teach the methods how to use the existing hooks. i.e.
> 
> ```python
> class Autoencoder(ModelMixin):
>     @use_forward_hooks
>     def decode(self, x):
>         ...
> ```
> 
> It's not perfect, because to implement the `use_forward_hooks` decorator you'd have to dig around in objects for state that there's no public API for (torch's `_forward_pre_hooks` and/or accelerate's `_hf_hook`). But we're pretty much ready to resort to that anyway.

Actually I really like this solution! Putting some more thought into this, I think it's also arguably bad design that we had the `encode()` and `decode()` functions in the first place. It's much more "PyTorchy" to only have `forward` methods like we have in `transformers`. 

In this case we should IMO have added the 4 lines of `decode` and `encode` directly to the pipelines and instead of passing one `vae` we could have passed `vae_encoder` and `vae_decoder` (this would have also allowed us to save memory better as the vae_encoder is actually never used in text-2-img. 

But it's too late now for this change -> think the nicest solution here would though to indeed create a (possibly complex) function decorator that "redirects" the hook from forward to encode or decode. Think that's def possible and since it's an edge case it's OK for me if this function decorator is a bit of black magic. 

Wdyt?

**Comment by patrickvonplaten** [comment]
@pcuenca I'd actually do the hacky thing right away in `diffusers` since it's as you said an edge case

File: src/diffusers/utils/accelerate_utils.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+from packaging import version
+
+from .import_utils import is_accelerate_available
+
+
+if is_accelerate_available():
+    import accelerate
+
+
+def apply_forward_hook(method):
```
Comment (patrickvonplaten) [inline src/diffusers/utils/accelerate_utils.py:27]: clean!

**Comment by patrickvonplaten** [comment]
Very cool! 

Could you maybe also:
- get `make quality` green
- add some docs to: https://huggingface.co/docs/diffusers/optimization/fp16
- add some tests 
- copy the functionality over to the most important other pipelines (maybe img2img, inpaint, pix2pix is enough)

**Comment by patrickvonplaten** [comment]
Looks great - thanks a lot! 

Can we maybe post some numbers on memory saving vs. speed-loss compared to original and "cpu_offload" ?

**Comment by Suhail** [comment]
Do you know why someone might get this error: `"LayerNormKernelImpl" not implemented for 'Half'` if they don't call pipe.enable_model_cpu_offload()

CleanShot 2023-03-05 at 11 15 55
CleanShot 2023-03-05 at 11 16 01

**Comment by pcuenca** [comment]
Hi @Suhail! Could be a device mismatch, but it's hard to tell. Was the text encoder moved `to("cuda")`, along with the rest of the pipeline? Can you please post a code snippet that we can use to reproduce the issue? Thanks!

**Comment by Suhail** [comment]
> 

Once I explicitly did:
```pipe = pipe.to('cuda')
pipe = pipe.to(torch.float16)```

Things seem to get fixed.

**Comment by yiyixuxu** [comment]
@Suhail 
the `LayerNormKernelImpl" not implemented for 'Half'` error happens when you run the pipeline on cpu so calling `to('cuda')` would make the error go away 

However we recommend use `pipe.enable_model_cpu_offload()` as it's more memory efficient - and if you call `pipe.enable_model_cpu_offload()`, you don't need to call `pipe.to('cuda')` afterward - 

https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/controlnet#diffusers.StableDiffusionControlNetPipeline.enable_model_cpu_offload

**Comment by Suhail** [comment]
Does anyone know how to do the control net mode conversion correctly or can point me to where the right directory to dl the necessary files?

I tried doing:
```
python scripts/convert_original_stable_diffusion_to_diffusers.py --controlnet --checkpoint_path ../models/control_sd15_canny.pth --dump_path ../models/control_sd15_canny --device cpu --original_config_file ./cldm_v15.yaml
```

CleanShot 2023-03-05 at 13 10 10

It worked but I only got this:
CleanShot 2023-03-05 at 13 11 01

**Comment by koren-v** [comment]
Hi, guys! 

I found memory leakage when using `enable_model_cpu_offload`, seems that something is not released... Do you have any ideas?
